ui/panel.js

```diff
@@ -0,0 +1,151 @@
+
+        this._hideIndicators();
+        this._updateBox(panel.left, this._leftBox);
+        this._updateBox(panel.center, this._centerBox);
+        this._updateBox(panel.right, this._rightBox);
+
+        if (panel.left.includes('dateMenu'))
+            Main.messageTray.bannerAlignment = Clutter.ActorAlign.START;
+        else if (panel.right.includes('dateMenu'))
+            Main.messageTray.bannerAlignment = Clutter.ActorAlign.END;
+        else
+            Main.messageTray.bannerAlignment = Clutter.ActorAlign.CENTER;
+
+        if (this._sessionStyle)
+            this._removeStyleClassName(this._sessionStyle);
+
+        this._sessionStyle = panelStyle;
+        if (this._sessionStyle)
+            this._addStyleClassName(this._sessionStyle);
+
+        if (this.get_text_direction() == Clutter.TextDirection.RTL) {
+            this._leftCorner.setStyleParent(this._rightBox);
+            this._rightCorner.setStyleParent(this._leftBox);
+        } else {
+            this._leftCorner.setStyleParent(this._leftBox);
+            this._rightCorner.setStyleParent(this._rightBox);
+        }
+    });
+
+    Utils.override(Panel.Panel, '_hideIndicators', function() {
+        const original = Utils.original(Panel.Panel, '_hideIndicators');
+
+        original.bind(this)();
+
+        for (let role in EXTRA_PANEL_ITEM_IMPLEMENTATIONS) {
+            let indicator = this.statusArea[role];
+            if (!indicator)
+                continue;
+            indicator.container.hide();
+        }
+    });
+
+    Utils.override(Panel.Panel, '_ensureIndicator', function(role) {
+        const original = Utils.original(Panel.Panel, '_ensureIndicator');
+
+        let indicator = original.bind(this)(role);
+        if (indicator)
+            return indicator;
+
+        let constructor = EXTRA_PANEL_ITEM_IMPLEMENTATIONS[role];
+        if (!constructor)
+            return null;
+
+        indicator = new constructor(this);
+        this.statusArea[role] = indicator;
+        _extraIndicators.push(indicator);
+
+        let destroyId = indicator.connect('destroy', emitter => {
+            let index = _extraIndicators.indexOf(indicator);
+            if (index > -1)
+                _extraIndicators.splice(index, 1);
+            emitter.disconnect(destroyId);
+        });
+
+        return indicator;
+    });
+
+    let panel = Main.panel;
+    // workaround to make sure the overriden Panel._updatePanel is
+    // invoked when the session mode is updated. The original signal
+    // connection uses 'bind' which circumvents the override mechanism here
+    _sessionModeUpdatedId = Main.sessionMode.connect('updated',
+        () => panel._updatePanel());
+    panel._updatePanel();
+}
+
+function disable() {
+    Utils.restore(Panel.Panel);
+
+    for (var i = _extraIndicators.length - 1; i >= 0; i--) {
+        let indicator = _extraIndicators[i];
+        indicator.destroy();
+    }
+    _extraIndicators = [];
+
+    if (_sessionModeUpdatedId) {
```

Comment:
`if (_sessionModeUpdateId > 0) {` for extra clarity

Reply:
Tbh I prefer to use `if (_sessionModeUpdatedId)` because it still works if we never set the variable - although not the case here this is the case in some indicators for example that only set these "source id" variables at a later point after construction (e.g. after a D-Bus proxy gets connected) and never set a default value for these variables on constructor.

That said I would prefer to leave it as is for consistency.